_posts/2021-09-24-01-serverless-rpc-rest-vs-rpc.md

```diff
@@ -0,0 +1,92 @@
+---
+layout: post
+title: Building a Serverless RPC API on AWS: REST vs RPC
+date: 2021-09-24
+tags: rest rpc grpc aws api
+author: Matt Tyler
+image: img/twirp.png
+---
+
+<center><img src="/img/twirp.png" /></center>
+<br/>
+
+There are many different ways to design an API but do any amount of searching and you will find that in 2021 there are three that dominate the discussion 
+
+- REST (REpresentational State Transfer),
+- RPC (Remote Procedure Call), and, 
+- GraphQL (Graph Query Language).
+
+They have an awful lot in common; 
+- There are many tools and frameworks that implement each,
+- You can find various examples of software companies implementing new APIs in each today, and,
+- JSON-over-HTTP is pretty commonly used for all three.
+
+If you want examples of each type. A large number of AWS APIs are implemented via RPC - DynamoDB being a good example of an RPC API. GitHub provides a fairly idiomatic REST API as well as a GraphQL API. Kubernetes makes fairly heavy use of RPC style APIs via GRPC.
+
+I recently spent a lot of time building a set of APIs to perform some infrastructure related tasks on AWS. In this case I went with what I’ve typically done which has included:
+
+- Designing the API using the OpenAPI document specification, and,
+- Using the specific to generate code for both the client and backend, and,
+- Using the server stubs to implement the API.
+
+The tooling in the OpenAPI ecosystem can be fairly inconsistent in terms of quality among different languages. I generally found the typescript tooling to be fairly lacking, nor was it documented particularly well. Despite this my team and I was able to make it work and paper over any cracks.
+
+But this did leave me wanting to explore what other options were out the in terms of designing, generating, and implementing APIs across different languages and frameworks.
+
+My first port of call was to inspect ["Smithy"](https://awslabs.github.io/smithy/index.html). Smith is an AWS IDL, similar in some ways to OpenAPI, which AWS is using to generate code for the newer AWS SDK’s. The v3 Javascript SDK and v2 Go SDK both use Smithy to generate code. Personally I find it a lot easier to read - similar to protobuf definitions, but a tad more pleasing to the human eye.
```

Comment:
I've been making sure external domain links open in a new tab, to keep your blog open. A reference can be found in the FED Talk series.